src/server/src/main.rs

```diff
@@ -1,23 +1,215 @@
+            async move 
+            {
+                loop 
+                {
+                    let msg = ctx.recv().await?;
+                    if let Ok(task) = msg.downcast::<String>() 
+                    {
+                        println!("Session for {} doing: {}", username_clone, task);
+                    }
+                    ctx.reply(format!("done {}", username_clone)).unwrap();
+                }
+                Ok(())
+            }
+        })
+    })
+    .expect("Failed to spawn user session")
+}
+
+async fn create_user_handler(
+    body: web::types::Json<User>,
+    path: web::types::Path<(String,)>,
+    user_map: web::types::Data<UserMap>,) -> impl Responder 
+{
+    let new_user = body.into_inner();
+    let (username,) = path.into_inner();
+    let mut map = user_map.lock().unwrap();
+
+    if map.contains_key(&username) {
+        return HttpResponse::build(418).body("User already exists.");
+    }
+
+    map.insert(username.clone(), new_user);
+    HttpResponse::Ok().body("User successfully created.")
+}
+
+async fn get_user_handler(
+    req: HttpRequest,
+    path: web::types::Path<(String,)>,
+    user_map: web::types::Data<UserMap>,
+    session_map: web::types::Data<SessionMap>,) -> impl Responder 
+{
+    let (username,) = path.into_inner();
+    let session_id = req.headers().get("SessionID").and_then(|val| val.to_str().ok());
+    let user_map = user_map.lock().unwrap();
+
+    if let Some(user) = user_map.get(&username) 
+    {
+        if let Some(session_id) = session_id
+        {
+            let session_map = session_map.lock().unwrap();
+            if let Some(sess_user) = session_map.get(session_id) 
+            {
+                // if current session matches. then return full data
+                println!("Session ID matched for user {}", username);
+                return HttpResponse::Ok().json(user);
+            } 
+        }
+        // else return public data 
+        let public_data = User 
+        {
+            username: user.username.clone(),
+            password: String::from("***"), 
+        };
+        HttpResponse::Ok().json(public_data)
+    } else 
+    {
+        HttpResponse::build(403).body("User does not exist.")
+    }
+}
+
+async fn update_user_handler(
+    req: HttpRequest,
+    path: web::types::Path<(String,)>,
+    update_value: web::types::Json<String>,
+    user_map: web::types::Data<UserMap>,
+    session_map: web::types::Data<SessionMap>,) -> impl Responder 
+{
+    let (username,) = path.into_inner();
+    let session_id = req.headers().get("SessionID").and_then(|val| val.to_str().ok());
+    let session_map = session_map.lock().unwrap();
+
+    if let Some(sess_user) = session_id.and_then(|sid| session_map.get(sid)) 
+    {
+        let mut user_map = user_map.lock().unwrap();
+        if let Some(user) = user_map.get_mut(&username) 
+        {
+            user.password = update_value.into_inner(); //whatever we want update to do, just pw rn
```

Comment:
yeah this has to be generalized. i wouldn't work with the concrete structures, because then any time you'd want to extend the `User` struct, you have to come back here (and a dozen other places) and remember to update them.
 i would just take a JSON payload (exactly like when we create a user), `into_inner()` it, and replace the `User` instance entirely with the new structure.